added masking to password

mahakporwal02 opened the pull request:
## Rationale
#525 
Passwords were visible to anonymous users
<!--
Issue: [Title](link) or #123 for Github issues.
-->
#525 
## Changes
Added masking to passwords(optimisation cache url and api key)
Passwords will only be visible to users with required permissions.
@rgaudin @kelson42 Pls decide which users should get the permission. At this point I have allowed permission to users with permission to update tasks.

rgaudin reviewed (changes requested):
thank you for this proposal but this won't work:
- your are removing this info for all users but those with schedules.update perm yet the main reason for those option to exist is to pass them to workers so they can use them in the scraper run
- you are using an hard-coded list of option names while those fields are already defined in the offliners def.
- protected value is still in the `str_command` field
- the way you search for the option is inefiscient and weak

ps: the import doesn't respect our code formatting

mahakporwal02 commented:
@rgaudin  In the above issue, masking is applied in the `scheduleview` so should the same thing be done in `taskvue`?
Because that would not actually solve the purpose as anyone could emulate the frontend and use your backend and get access to the passwords hidden as `secret replacement key`
If we do it according to my last PR we would face problem with giving access to the `worker`, so if there is a way to give access to the worker the main problem would be solved?

rgaudin commented:
> @rgaudin In the above issue, masking is applied in the `scheduleview` so should the same thing be done in `taskvue`?

I don't get that part. Your PR only modifies `task.py` so that only applies to the JSON task API

> Because that would not actually solve the purpose as anyone could emulate the frontend and use your backend and get access to the passwords hidden as `secret replacement key`

That's the current situation. Password are hidden in the GUI but are available via the JSON API. I think the ticket is to prevent this as well… yet allowing both workers and schedule editors to access it.

> If we do it according to this PR we would face problem with giving access to the `worker`, so if there is a way to give access to the worker so the main problem would be solved?

Yes. Workers have a different set of permissions so you could use that. I am not a fan of having endpoints respond differently based on who's requesting – and actually I don't think OpenAPI allows us to describe this– so we should check with @kelson42 about the way forward:
- respond differently to `/tasks/xxx` based on the user making the request
- add a new endpoint like `tasks/xxx/public` that the UI for unauthenticated users would use

I'm in favor of the latter but that's a bit more work.

mahakporwal02 commented:
> I don't get that part. Your PR only modifies `task.py` so that only applies to the JSON task API
> 
> >

Yes I have only modified `task.py` but I was talking about masking already been done on `scheduleView`
![Screenshot from 2021-02-25 15-43-48](https://user-images.githubusercontent.com/56486682/109138430-5cb7ca00-7780-11eb-944e-1e5f43799c5a.png)

rgaudin commented:
> > I don't get that part. Your PR only modifies `task.py` so that only applies to the JSON task API
> > >
> 
> Yes I have only modified `task.py` but I was talking about masking already been done on `scheduleView`

OK ; I think I responded in the rest of my comment then

kelson42 commented:
I confirm that workers and editors/admins (all roles with write permissions) should get the same API JSON content. All others should get offuscated passwords (with star characters). This should be documented in API documentation each time it is necessary.

mahakporwal02 commented:
@rgaudin So now imo 
1) I should add permissions for the wrokers' access.
2) Add a more efficient method for searching options.
3) Hide passwords in `str_command` also.
4) Remove hard coded names and instead add definitions from `offliner_def`.?

mahakporwal02 commented:
@rgaudin So now imo

1) I should add permissions for the `workers'` access.
2) Add a more efficient method for searching options.
3) Hide passwords in `str_command` also.
4) Remove hard coded names and instead add definitions from `offliner_def`.?

rgaudin commented:
> @rgaudin So now imo
>     1. I should add permissions for the `workers'` access.

Yes and No. Workers don't access schedule endpoint. Workers access the task endpoint.
You should make your code resusable and use it at both locations:
- in task only for those with `tasks.create` perm
- in schedule only for those with `schedules.create` or `schedules.update` perm
 
>     2. Add a more efficient method for searching options.

Yes.

>     3. Hide passwords in `str_command` also.

Yes

>     4. Remove hard coded names and instead add definitions from `offliner_def`.?

Right. Offliner definitions (in `schemas/offliners/`) have fields for which there's a `secret=True` metadata. That's what you want to apply your filter on.

kelson42 commented:
@mahakporwal02 Please rebase your PR on master.

mahakporwal02 commented:
@kelson42 @rgaudin I have rebased the code, code factor fails as I have added `******` for secret key as a constant, So what should be done to pass it?

rgaudin reviewed (changes requested):
Thank you ; this works as expected.
I've indicated a number of small changes in the python code.

For the UI, I think we should keep the information that such a field contains a secret so that the user is informed that's it's not a proper `*****` use. Either just a tooltip or an icon should do it.

Also, you didn't update the openapi documentation to inform about this new behavior…

mahakporwal02 commented on dispatcher/backend/src/common/utils.py:
```diff
@@ -334,3 +336,29 @@ def handle_others(self, event):
     logger.info(f"Other event, keys: {list(event.keys())}")
+
+
+def hide_secret_flags(response):
```
Do you mean that I should create a file named `utils` in `routes`  and paste the above code to this file?
and import the function from the newly created file in `task.py` and `schedule.py`?

mahakporwal02 commented:
> Thank you ; this works as expected.
> I've indicated a number of small changes in the python code.
> 
> For the UI, I think we should keep the information that such a field contains a secret so that the user is informed that's it's not a proper `*****` use. Either just a tooltip or an icon should do it.
> 
> Also, you didn't update the openapi documentation to inform about this new behavior…

In openapi documentation, should I just change the `discription` to inform about the response to unautorized users?

rgaudin commented on dispatcher/backend/src/common/utils.py:
```diff
@@ -334,3 +336,29 @@ def handle_others(self, event):
+                    response["container"]["command"][index] = (
+                        "--" + secret_field + "=" + SECRET_REPLACEMENT
+                    )
+        response["config"]["str_command"] = " ".join(response["config"]["command"])
```
The function should be in offliners.py but it should only be called from your new function (and where that piece of code was used in offliners)

rgaudin commented:
> In openapi documentation, should I just change the `discription` to inform about the response to unautorized users?

Yep, it think that would be enough

rgaudin reviewed (changes requested):
Couple of very minor stuff left ; we're almost there!

mahakporwal02 commented on dispatcher/backend/docs/openapi_v1.yaml:
```diff
@@ -309,7 +309,7 @@ paths:
       summary: list schedules
       operationId: getSchedule
-      description: Retrieve a Schedule's details
+      description: Retrieve a Schedule's details (returns ****** for secret keys for unauthorized users)
```
Retrieve a Schedule's details (returns ****** for secret keys for users without required permissions)
will do?

rgaudin replied:
```
Retrieve a Schedule's details (Secrets in config are replaced by ****** for unauthenticated requests and authenticated requests from users without schedules.update permission)
```

IN task, you'd change the perm to tasks.create of course

rgaudin reviewed (approved):
Thank you @mahakporwal02!
